### PR TITLE
Make zstd convert safe to use concurrently

### LINKED ISSUE #3078: ZstdLayerConvertFunc unsafe to use concurrently
### Description

`convert zstd` is not safe to use concurrently, as there is no guarantee that the temporary uncompressed resource is still there by the time we read it.

https://github.com/containerd/nerdctl/blob/main/pkg/imgutil/converter/zstd.go#L54

Furthermore, using `uncompress.LayerConvertFunc` seems to create useless churn (and is at the root of the problem), as we will read and write twice instead of once.

From a quick read, it seems to me that we should instead just avoid the containerd Uncompress helper altogether and just get a DecompressStream directly and roll with it:

```golang

	readerAt, err := cs.ReaderAt(ctx, desc)
	sr := io.NewSectionReader(readerAt, 0, desc.Size)
	newR, err := compression.DecompressStream(sr)

```

Thoughts?

### Steps to reproduce the issue

Use convert zstd concurrently.

### Describe the results you received and expected

Fail because the temporary uncompressed layer has already been deleted by the concurrent execution.

Should not fail.

### What version of nerdctl are you using?

1.7.6

### Are you using a variant of nerdctl? (e.g., Rancher Desktop)

None

### Host information

_No response_

### PR DESCRIPTION
**Pull request by apostasie**
Fix #3078 by removing use of intermediary uncompressed layer.

This incidentally gives us a small (?) perf boost (as we no longer read and write resources twice).

This will make convert tests able to be run concurrently.

@AkihiroSuda and team - PTAL at your convenience

### CODE REVIEW

**Comment by apostasie** [comment]
About the "perf boost": un-scientific measurement show 30% speed-up at minimum compression for Alpine - and ~15% at max compression.

File: pkg/imgutil/converter/zstd.go [hunk 1]
```diff
@@ -40,34 +41,35 @@ func ZstdLayerConvertFunc(options types.ImageConvertOptions) (converter.ConvertF
...
+		var oldReader io.Reader
+		// If it is compressed, get a decompressed stream
+		if !uncompress.IsUncompressedType(desc.MediaType) {
+			decompStream, err := compression.DecompressStream(sectionReader)
+			if err != nil {
+				return nil, err
+			}
+			defer decompStream.Close()
+			oldReader = decompStream
+			delete(labelsMap, labels.LabelUncompressed)
```
Comment (apostasie) [inline pkg/imgutil/converter/zstd.go:68]: Note that I am not quite sure I understand why containerd is doing that in the helper - but I wanted to make sure we have iso behavior.

**Comment by apostasie** [comment]
Adding cleaned-up test as well (now that we can parallelize).

File: pkg/imgutil/converter/zstd.go (hunk 1 above)
Comment (ktock) [reply pkg/imgutil/converter/zstd.go:68]: Please do not delete this but update this value if the actual uncompressed digest differs from the original one. 

This is defined at: https://github.com/containerd/containerd/blob/4a18adcfca4e37e829c0d3174bd04d4cbf222f9a/pkg/labels/labels.go#L21
This is used several use cases for example diff computation: https://github.com/containerd/containerd/blob/4a18adcfca4e37e829c0d3174bd04d4cbf222f9a/core/images/diffid.go#L47 
Other clients also might use it.

Comment (apostasie) [reply pkg/imgutil/converter/zstd.go]: Ok.

Curious what your take is on the following in that context - as I said, not sure what there is to gain in doing it containerd?
https://github.com/containerd/containerd/blob/v1.7.18/images/converter/uncompress/uncompress.go#L80

Anyhow, re-pushed.

**Comment by apostasie** [comment]
> [#3079 (comment)](https://github.com/containerd/nerdctl/pull/3079#discussion_r1634447937)

Fixed.

**Comment by ktock** [comment]
Could you update labels.LabelUncompressed with the actual uncompressed digest, using Digester of go-digest? e.g.: https://github.com/containerd/containerd/blob/4a18adcfca4e37e829c0d3174bd04d4cbf222f9a/core/diff/apply/apply.go#L90

**Comment by ktock** [comment]
> Can you clarify why we should change this behavior?
> Is the current logic faulty on that?

The current logic isn't faulty but it's good to make sure that the content has the correct LabelUncompressed (this can be used for fast path as mentioned in https://github.com/containerd/nerdctl/pull/3079#discussion_r1634447937 ). But yes, we can do that as a following up so this patch looks good for now.

**Comment by apostasie** [comment]
Thanks a lot @ktock 
Appreciate the fast review on this!

**Pull request merged**